File: Midterm_EquipmentRental_Team5/Controllers/EquipmentController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Midterm_EquipmentRental_Team5.Models;
using Midterm_EquipmentRental_Team5.Models.Interfaces;
using Midterm_EquipmentRental_Team5.Services;
using Midterm_EquipmentRental_Team5.Services.Interfaces;

namespace Midterm_EquipmentRental_Team5.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    [Authorize] // All endpoints require authentication
    public class EquipmentController(IEquipmentServices equipmentService, IWebHostEnvironment environment) : ControllerBase
    {
        private readonly IEquipmentServices _equipmentService = equipmentService;
        // GET /api/equipment - Get all equipment with pagination
        [HttpGet]
        public ActionResult<IEnumerable<IEquipment>> GetAllEquipment(int page = 1)
        {
            try
            {
                var equipment = equipmentService.GetAllEquipment(page);
                return Ok(equipment);
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(ex.Message);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        // GET /api/equipment/{id} - Get specific equipment details
        [HttpGet("{id}")]
        public ActionResult<IEquipment> GetEquipment(int id)
        {
            try
            {
                var equipment = _equipmentService.GetEquipmentById(id);

                if (equipment == null)
                {
                    return NotFound($"Equipment with ID {id} not found.");
                }

                return Ok(equipment);
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(ex.Message);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        // POST /api/equipment - Add new equipment (Admin only)
        [HttpPost]
        [Authorize(Roles = "Admin")]
        public ActionResult AddEquipment([FromBody] Equipment newEquipment)
        {
            try
            {
               _equipmentService.AddEquipment(newEquipment);
                return NoContent();
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        // PUT /api/equipment/{id} - Update equipment (Admin only)
        [HttpPut("{id}")]
        [Authorize(Roles = "Admin")]
        public ActionResult UpdateEquipment(int id, [FromBody] Equipment updatedEquipment)
        {
            try
            {
                _equipmentService.UpdateEquipment(id, updatedEquipment);
                return NoContent();
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(ex.Message);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        // DELETE /api/equipment/{id} - Delete equipment (Admin only)
        [HttpDelete("{id}")]
        [Authorize(Roles = "Admin")]
        public ActionResult DeleteEquipment(int id)
        {
            try
            {
                _equipmentService.DeleteEquipment(id);
                return NoContent();
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(ex.Message);
            }

            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        // GET /api/equipment/available - Get list of available equipment
        [HttpGet("available")]
        public ActionResult<IEnumerable<IEquipment>> GetAvailableEquipment()
        {
            try
            {
                var availableEquipment = _equipmentService.GetAvailableEquipment();
                return Ok(availableEquipment);
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(ex.Message);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }
        // GET /api/equipment/rented - Get rented equipment summary (Admin only)
        [HttpGet("rented")]
        [Authorize(Roles = "Admin")]
        public ActionResult<IEquipment> GetRentedEquipmentSummary()
        {
            try
            {
                var rentedEquipment = _equipmentService.GetRentedEquipment();
                return Ok(rentedEquipment);
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(ex.Message);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }
    }
}

File: Midterm_EquipmentRental_Team5_FrontEnd/src/api/EquipmentController.js
import RequestHandler from '@/services/RequestHandler'

export async function getAllEquipment(page = 1) {
  try {
    const res = await RequestHandler.get(`equipment?page=${page}`)
    return res
  } catch (error) {
    console.error('Error fetching all equipment:', error)
  }
}

export async function getEquipment(id) {
  try {
    const res = await RequestHandler.get(`equipment/${id}`)
    return res
  } catch (error) {
    console.error(`Error fetching equipment with ID ${id}:`, error)
  }
}

export async function addEquipment(newEquipment) {
  try {
    const res = await RequestHandler.post('equipment', newEquipment)
    return res
  } catch (error) {
    console.error('Error adding new equipment:', error)
  }
}

export async function getEquipmentRentalHistory(equipmentId) {
  try {
    const res = await RequestHandler.get(`equipment/${equipmentId}/rental-history`)
    return res
  } catch (error) {
    console.error(`Error fetching rental history for equipment ID ${equipmentId}:`, error)
    throw error
  }
}

export async function updateEquipment(id, updatedEquipment) {
  try {
    const res = await RequestHandler.put(`equipment/${id}`, updatedEquipment)
    return res
  } catch (error) {
    console.error(`Error updating equipment with ID ${id}:`, error)
  }
}

export async function deleteEquipment(id) {
  try {
    const res = await RequestHandler.delete(`equipment/${id}`)
    return res
  } catch (error) {
    console.error(`Error deleting equipment with ID ${id}:`, error)
  }
}

export async function getAvailableEquipment() {
  try {
    const res = await RequestHandler.get('equipment/available')
    return res
  } catch (error) {
    console.error('Error fetching available equipment:', error)
  }
}

export async function getRentedEquipmentSummary() {
  try {
    const res = await RequestHandler.get('equipment/rented')
    return res
  } catch (error) {
    console.error('Error fetching rented equipment summary:', error)
  }
}
